Document non-pytorch CUDA memory allocation and how to query it

wconstab opened the pull request:
This PR documents the fact that PyTorch does not have visibility into how every CUDA memory allocation happend - it only knows about allocations that went through the pytorch CUDA allocator.

It also adds a code snippet showing how to use pynvml to query current GPU memory usage.


## Preview
Added a note at the top of "Understanding CUDA Memory Usage" doc:
<img width="732" alt="image" src="https://github.com/user-attachments/assets/69e28d2a-841a-4b1b-b886-e96fb5d76582" />

which links to a section below:
<img width="733" alt="image" src="https://github.com/user-attachments/assets/cab4f252-9ac2-4fc6-a45d-fdb958fc7dbc" />


   
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150880

wconstab committed:
WIP

[ghstack-poisoned]

wconstab committed:
Update on "WIP"




[ghstack-poisoned]

wconstab committed:
Update on "WIP"




[ghstack-poisoned]

wconstab committed:
Update on "WIP"




[ghstack-poisoned]

wconstab committed:
Update on "Document non-pytorch CUDA memory allocation and how to query it"

This PR documents the fact that PyTorch does not have visibility into how every CUDA memory allocation happend - it only knows about allocations that went through the pytorch CUDA allocator.

It also adds a code snippet showing how to use pynvml to query current GPU memory usage.
   



[ghstack-poisoned]

kwen2501 commented on docs/source/torch_cuda_memory.rst:
```diff
@@ -8,6 +8,15 @@ at any point in time, and optionally record the history of allocation events tha
+   allocted and managed through the PyTorch allocator.  Any memory allocated directly from CUDA APIs will not be
+   visible in the PyTorch memory profiler.
+
+   NCCL (used for distributed communication on CUDA devices) is a common example of a library that allocates significant
```
Shall we change "significant" to some word softer?

wconstab replied:
nontrivial?

kwen2501 reviewed (approved):
Thanks for the documentation. I added one comment.

wconstab committed:
Update on "Document non-pytorch CUDA memory allocation and how to query it"

This PR documents the fact that PyTorch does not have visibility into how every CUDA memory allocation happend - it only knows about allocations that went through the pytorch CUDA allocator.

It also adds a code snippet showing how to use pynvml to query current GPU memory usage.
   



[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

ngimel commented on docs/source/torch_cuda_memory.rst:
```diff
@@ -54,6 +63,30 @@ The address b7f064c000000_0 refers to the (b)lock at address 7f064c000000 which
 
+.. _non_pytorch_alloc:
+
+Identifying Non-PyTorch allocations
```
we already have a util that does this, `torch.cuda.device_memory_used` that also correctly handles device reordering that might happen with CUDA_VISIBLE_DEVICES, which won't work with naive application of the snippet below

wconstab replied:
TIL! ok, i will update to reference that instead.  

IIUC your point is that if i pass '0' to pynvml, it will spawn a new process that escapes CUDA_VISIBLE_DEVICES and it may see a different device0 as the current torch process sees?

wconstab replied:
btw is this doc correct?  is .memory_usage() giving us stats about memory bw utilization? or is it actually reporting the current allocation?

<img width="1106" alt="image" src="https://github.com/user-attachments/assets/79a33f36-5a83-44c2-8165-4b708a3e7ae1" />

stas00 replied:
`CUDA_DEVICE_ORDER` can impact this as well. please see:

https://github.com/stas00/ipyexperiments/blob/569a450e204b1da60e5ef07a96c91553b286ea14/docs/ipyexperiments.md#different-ordering-of-gpu-ids

ngimel replied:
Yeah we should be doing this based on uuid, and not on parsing env vars :-(

ngimel replied:
pynvml doesn't spawn new processes, but when doing `pynvml.nvmlDeviceGetHandleByIndex(device_idx)` it will ignore whatever reordering runtime was seeing due to CUDA_DEVICE_ORDER and CUDA_VISIBLE_DEVICES and just go with pci order. In any case, using pytorch utilities is better than raw pynvml code that you have here.

wconstab replied:
updated it to show the torch API.

stas00 commented on docs/source/torch_cuda_memory.rst:
```diff
@@ -8,6 +8,15 @@ at any point in time, and optionally record the history of allocation events tha
+.. note::
+
+   The memory profiler and visualizer described in this document only have visibility into the CUDA memory that is
+   allocted and managed through the PyTorch allocator.  Any memory allocated directly from CUDA APIs will not be
```
```suggestion
   allocated and managed through the PyTorch allocator.  Any memory allocated directly from CUDA APIs will not be
```

wconstab committed:
Update on "Document non-pytorch CUDA memory allocation and how to query it"

This PR documents the fact that PyTorch does not have visibility into how every CUDA memory allocation happend - it only knows about allocations that went through the pytorch CUDA allocator.

It also adds a code snippet showing how to use pynvml to query current GPU memory usage.


## Preview
Added a note at the top of "Understanding CUDA Memory Usage" doc:
<img width="732" alt="image" src="https://github.com/user-attachments/assets/69e28d2a-841a-4b1b-b886-e96fb5d76582" />

which links to a section below:
<img width="733" alt="image" src="https://github.com/user-attachments/assets/cab4f252-9ac2-4fc6-a45d-fdb958fc7dbc" />


   



[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
